website/events-listener/src/kafka/producer.ts

```diff
@@ -1,71 +1,62 @@
+
+async function connect(): Promise<void> {
+  await producer.connect();
+  eventListenerLogger.info('Producer is connected');
+}
 
-export class KafkaProducer {
-  readonly kafka: Kafka;
-  readonly producer: Producer;
+async function send(key: string, value: string, genesis: string): Promise<void> {
+  await producer.send({
+    topic: 'events',
+    messages: [{ key, value: JSON.stringify(value), headers: { genesis } }],
+  });
+}
+
+async function createTopic(topic: string): Promise<void> {
+  const admin = initKafka.admin();
+  await connectKafkaAdmin(admin);
+  await createKafkaTopic(admin, topic);
+  await admin.disconnect();
+  eventListenerLogger.info(`Admin is disconnected`);
+}
 
-  constructor() {
-    this.kafka = new Kafka({
-      clientId: config.kafka.clientId,
-      brokers: config.kafka.brokers,
-      logCreator: kafkaLogger,
-      sasl: {
-        mechanism: 'plain',
-        username: config.kafka.sasl.username,
-        password: config.kafka.sasl.password,
-      },
-    });
-    this.producer = this.kafka.producer();
+async function connectKafkaAdmin(admin: Admin): Promise<void> {
+  try {
+    await admin.connect();
+    eventListenerLogger.info('Kafka producer: Admin is connected');
+  } catch (error) {
+    eventListenerLogger.error(error);
+    eventListenerLogger.error('Admin is not connected');
+    throw error;
   }
+}
 
-  async createTopic(topic: string) {
-    const admin = this.kafka.admin();
-    try {
-      await admin.connect();
-      eventListenerLogger.info('Kafka producer: Admin is connected');
-    } catch (error) {
-      eventListenerLogger.error(error);
-      eventListenerLogger.error('Admin is not connected');
-      throw error;
-    }
-    try {
-      const topics = await admin.listTopics();
-      if (!topics.includes(topic)) {
-        await admin.createTopics({
-          waitForLeaders: true,
-          topics: [
-            {
-              topic,
-            },
-          ],
-        });
-        eventListenerLogger.info(`Kafka producer: Topic ${topic} created`);
-      } else {
-        eventListenerLogger.warn(`Kafka producer: Topic ${topic} already existed`);
-      }
-    } catch (error) {
-      eventListenerLogger.error(`Kafka producer: ${error}`);
-      await admin.disconnect();
-      eventListenerLogger.info(`Admin is disconnected`);
-      throw error;
+async function createKafkaTopic(admin: Admin, topic: string): Promise<void> {
+  try {
+    const kafkaTopics = await admin.listTopics();
+    if (!isTopicAlreadyExist(kafkaTopics, topic)) {
+      await admin.createTopics({
+        waitForLeaders: true,
+        topics: [{ topic }],
+      });
+      eventListenerLogger.info(`Kafka producer: Topic ${topic} created`);
+    } else {
+      eventListenerLogger.warn(`Kafka producer: Topic ${topic} already existed`);
     }
+  } catch (error) {
+    eventListenerLogger.error(`Kafka producer: ${error}`);
     await admin.disconnect();
```

Comment:
`admin.disconnected()` should be called in any case, not only in case of an error

Reply:
@osipov-mit  Do you mean like this way ?

```javascript
async function createKafkaTopic(admin: Admin, topic: string): Promise<void> {
  try {
    const kafkaTopics = await admin.listTopics();
    if (!isTopicAlreadyExist(kafkaTopics, topic)) {
      await admin.createTopics({
        waitForLeaders: true,
        topics: [{ topic }],
      });
      eventListenerLogger.info(`Kafka producer: Topic ${topic} created`);
    } else {
      eventListenerLogger.info(`Kafka producer: Topic ${topic} already existed`);
    }
  } catch (error) {
    eventListenerLogger.error(`Kafka producer: ${error}`);
    eventListenerLogger.info(`Admin is disconnected`);
    throw error;
  }
  
  await admin.disconnect();
}
```

Reply:
> @osipov-mit `amin.disconnected()` was there before in `catch` area. I didn't change it. In `master` branch currently the same in `catch` scope.

No, it was in catch and in after try/catch area. But I think there is no need to keep it in catch area. Because in that case application will go down and admin will be disconnected anyway

Reply:
I mean
```javascript
async function createKafkaTopic(admin: Admin, topic: string): Promise<void> {
  try {
    const kafkaTopics = await admin.listTopics();
    if (!isTopicAlreadyExist(kafkaTopics, topic)) {
      await admin.createTopics({
        waitForLeaders: true,
        topics: [{ topic }],
      });
      eventListenerLogger.info(`Kafka producer: Topic ${topic} created`);
    } else {
      eventListenerLogger.info(`Kafka producer: Topic ${topic} already existed`);
    }
  } catch (error) {
    eventListenerLogger.error(`Kafka producer: ${error}`);
    throw error;
  }
  
  await admin.disconnect();
  eventListenerLogger.info(`Admin is disconnected`);
}
```